### ISSUE TITLE
How to escape AMP query parameter right?

### ISSUE THREAD

**Issue by danil-smirnov** [okigan/awscurl#169]
Hi,

I'm trying to query AMP instance with` awscurl` like this:

`docker run --rm -it okigan/awscurl -i --access_key $AWS_ACCESS_KEY_ID --secret_key $AWS_SECRET_ACCESS_KEY --session_token $AWS_SESSION_TOKEN --region eu-central-1 --service aps $AMP_QUERY_ENDPOINT'?query=http_request_duration_seconds_bucket{}'`

It works fine until I add a parameter between curly brackets:

`docker run --rm -it okigan/awscurl -i --access_key $AWS_ACCESS_KEY_ID --secret_key $AWS_SECRET_ACCESS_KEY --session_token $AWS_SESSION_TOKEN --region eu-central-1 --service aps $AMP_QUERY_ENDPOINT'?query=http_request_duration_seconds_bucket{status="2xx"}'`

I'm getting `InvalidQueryStringException` error in the latter case, though it works fine in the Grafana preview.

I tried different escape methods but haven't found a working one.

**Comment by okigan** [comment]
try to escape the parameters (you could use https://www.urlencoder.org/),
so something like this:

docker run --rm -it okigan/awscurl -i --access_key "$AWS_ACCESS_KEY_ID"
--secret_key "$AWS_SECRET_ACCESS_KEY" --session_token "$AWS_SESSION_TOKEN"
--region eu-central-1 --service aps
"${AMP_QUERY_ENDPOINT}?query=http_request_duration_seconds_bucket%7Bstatus%3D%222xx%22%7D"

**Comment by danil-smirnov** [comment]
The error is the same if I do

`docker run --rm -it okigan/awscurl -i --access_key "$AWS_ACCESS_KEY_ID" --secret_key "$AWS_SECRET_ACCESS_KEY" --session_token "$AWS_SESSION_TOKEN" --region eu-central-1 --service aps "${AMP_QUERY_ENDPOINT}?query%3Dhttp_request_duration_seconds_bucket%7Bstatus%3D%222xx%22%7D"`

I saw some closed issues mentioning this, I suppose we have a bug here

**Comment by okigan** [comment]
Please add link(s) to relevant ones here - thanks!

**Comment by danil-smirnov** [comment]
@okigan Here it is:

https://github.com/okigan/awscurl/issues?q=is%3Aissue+escape

**Comment by okigan** [comment]
@danil-smirnov This is a tricky issue, especially for existing users...

Could you confirm source in this PR resolve your issue (check out the PR branch) and test similarly as shown in the top comment, but with AMP query: https://github.com/okigan/awscurl/pull/171#issue-1708219549

**Comment by danil-smirnov** [comment]
@okigan I've build a Docker image from branch `issue/169` and tried all the variants but I still got `InvalidQueryStringException` or  `InvalidSignatureException` errors

**Comment by alediaferia** [comment]
Is there any update for this or any recommended alternative approach?

**Comment by okigan** [comment]
Yeah with the PR above we (@danil-smirnov and I) were able to get it to work. 

@alediaferia are you using branch build? with/without docker?

**Comment by yuwuc** [comment]
@okigan I think the issue is in this [line](https://github.com/okigan/awscurl/blob/master/awscurl/awscurl.py#L334). Double url encode. After removing function `aws_url_encode()`, my prometheus query works.

Sample command:

`awscurl -X POST --region us-west-2 --service aps "${AMP_QUERY_ENDPOINT}query?query=group%20by%28__name__%29%20%28%7B__name__%21%3D%22%22%7D%29"`

You could see the query string changed after enabling debugging:
`('\n'
 'CANONICAL REQUEST = POST\n'
 '/workspaces/-----/api/v1/query\n'
'query=group%2520by%2528__name__%2529%2520%2528%257B__name__%2521%253D%2522%2522%257D%2529\n'
 'host:aps-workspaces.us-west-2.amazonaws.com\n'
 'x-amz-date:20230803T200239Z\n'`

The query string changed from `group%20by%28__name__%29%20%28%7B__name__%21%3D%22%22%7D%29` to `group%2520by%2528__name__%2529%2520%2528%257B__name__%2521%253D%2522%2522%257D%2529`

**Comment by okigan** [comment]
Yes the issue is about encoding, question is who suppose to do it. In your example you already encoded them them before calling the awscurl but some users don’t. I’ve reread this article [1] by maker of curl, but did not arrive at succinct conclusion. Give it a shot, tell me what make sense (with reference). [1] https://daniel.haxx.se/blog/2016/05/11/my-url-isnt-your-url/On Aug 3, 2023, at 3:41 PM, Yu-Wu Chu ***@***.***> wrote:﻿
@okigan I think the issue is in this line. Double url encode. After removing function aws_url_encode(), my prometheus query works.
Sample command:
awscurl -X POST --region us-west-2 --service aps "${AMP_QUERY_ENDPOINT}query?query=group%20by%28__name__%29%20%28%7B__name__%21%3D%22%22%7D%29"
You could see the query string changed after enabling debugging:
('\n' 'CANONICAL REQUEST = POST\n' '/workspaces/-----/api/v1/query\n' 'query=group%2520by%2528__name__%2529%2520%2528%257B__name__%2521%253D%2522%2522%257D%2529\n' 'host:aps-workspaces.us-west-2.amazonaws.com\n' 'x-amz-date:20230803T200239Z\n'
The query string changed from group%20by%28__name__%29%20%28%7B__name__%21%3D%22%22%7D%29 to group%2520by%2528__name__%2529%2520%2528%257B__name__%2521%253D%2522%2522%257D%2529

**Comment by yuwuc** [comment]
@okigan Thanks for sharing the article. It is interesting. You are right. This might need to think more about it.

**Comment by yuwuc** [comment]
I found another work around. Hopefully, this helps who faces this issue via POST and move query string to body.

`awscurl -X POST --region us-west-2 --service aps "${AMP_QUERY_ENDPOINT}/query" -d 'query=group by(__name__) ({__name__!=""})' --header 'Content-Type: application/x-www-form-urlencoded'`

**Comment by okigan** [comment]
I think that's a great find! -- side steps this whole issue
